### DISCUSSION TITLE
kubi | a cubemap generator

### DISCUSSION THREAD

**Discussion by indus** [libvips/libvips#2299] [Show and tell]
I made a cubemap generator with pyvips: **[kubi](https://github.com/indus/kubi)**
Like everything else I've made with vips so far it is quite fast 🚀 

Motivation: I had the need to transform thousands of equirectangular frames to relatively large cubemaps with a special transform function to render an animation in the end. I din't wan't to wait a full day for the cubemap generation, so I invested not one but two full days :stuck_out_tongue_winking_eye: to come up with **kubi**. Now it runs in a couple of hours. 

I developed this on Windows with limited python experience.
Feedback from someone with more python knowledge or a test on Linux would be welcome.

**Comment by jcupitt** [comment]
Oh that's very cool! Nice work.

Your fiddling with `PATH` should probably be behind a platform test, eg.:

```python
import platform

if platform.system() == 'Windows':
    os.envion['PATH'] = ...
```

You could try copying the index image to memory before looping over images, eg.:

```python
index = index.copy_memory()
```

That'll render the whole index image to a large memory array, rather than relying in delayed computation. It'll mean you only pay index construction cost once, rather than each time you use it. Perhaps the speedup wouldn't be very useful.

**Reply by jcupitt** [reply]
Oh, and the numpy can probably be reimplemented with `xyz` or maybe `grey`, but again, it's probably not very important for speed.

**Reply by indus** [reply]
Thanks for your feedback.

The hardcoded PATH never should have made it to the release. I was just tired in using kubis ```--vips``` argument and forgot to delete it. It is now removed and ```--vips``` is also limited to Windows users.

I made a test with ```copy_memory```: https://github.com/indus/kubi/blob/main/src/kubi/kubi.py#L170-L175
But I havn't found any changes in performance. Maybe because the index at this point only holds float values between 0 and 1. It gets scaled for every image separately to allow varying input dimensions. I will have to look into this further.

I really tried to replace numpy with vips and looked into ```xyz``` and ```polar``` as I realized no ```atan2``` exists in vips. For now I havn't found a solution to replace the numpy code block. In another experiment I got noticible performance improvement by replacing a  ```np.linspace```+```np.meshgrid``` with ```vips.xyz```. But because all the numpy code only runs once to create the index and not in the loop it is not that important. Just the idea of 100% vips  would be nice :grin: 


I saw that [VipsOperationMath2](https://libvips.github.io/libvips/API/current/libvips-arithmetic.html#VipsOperationMath2) is not that rich in functions. Would you accept a PR for ```VIPS_OPERATION_MATH2_ATAN2 | atan2 ( left, right )```?

**Reply by indus** [reply]
There is already this check https://github.com/libvips/libvips/blob/master/configure.ac#L399 🤔

**Reply by jcupitt** [reply]
Yes, that's used for `polar`. Sure, I can see an atan2 would be useful. There's `vips_complex_atan2()` in complex.c you can reuse / adapt etc.
